server/trillian_log_server.go

```diff
@@ -75,26 +75,30 @@ func (t *TrillianLogServer) GetSequencedLeafCount(ctx context.Context, req *tril
 }
 
 func (t *TrillianLogServer) GetLeavesByIndex(ctx context.Context, req *trillian.GetLeavesByIndexRequest) (*trillian.GetLeavesByIndexResponse, error) {
+	if !validateLeafIndices(req.LeafIndex) {
+		return &trillian.GetLeavesByIndexResponse{Status: buildStatusWithDesc(trillian.TrillianApiStatusCode_ERROR, "Invalid -ve leaf index in request")}, nil
+	}
+
 	tx, err := t.prepareStorageTx(*req.LogId)
 
 	if err != nil {
 		return nil, err
 	}
 
-	if !validateLeafIndices(req.LeafIndex) {
-		return &trillian.GetLeavesByIndexResponse{Status: buildStatusWithDesc(trillian.TrillianApiStatusCode_ERROR, "Invalid -ve leaf index in request")}, nil
-	}
-
-	// Read only transaction, failure to commit not propagated to client
-	defer t.commitAndLog(tx, "GetLeavesByIndex")
-
 	leaves, err := tx.GetLeavesByIndex(req.LeafIndex)
 
 	if err != nil {
+		tx.Rollback()
 		return nil, err
 	}
 
 	leafProtos := leavesToProtos(leaves)
+
+	if err := t.commitAndLog(tx, "GetLeavesByIndex"); err != nil {
+		tx.Rollback()
```

Comment:
Actually we could define the API so the TX impl must clean up if the commit fails then not worry about it. Is that better?